### PR TITLE
Integration server

### PR DESCRIPTION
**Pull request by hendrikebbers**
New module that provides a server app to test the DP with Docker.

---
This change is [Reviewable](https://reviewable.io/reviews/canoo/dolphin-platform/399)

### CODE REVIEW

**Comment by hendrikebbers** [comment]
Currently the test are mostly working (1 test is broken based on issue in Calendar conversion)

**Comment by hendrikebbers** [comment]
Sadly the Gradle build is running in the wrong order

File: integration-common/src/main/java/com/canoo/dolphin/integration/IntegrationConstants.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.canoo.dolphin.integration;
+
+/**
```
Comment (timo0) [inline integration-common/src/main/java/com/canoo/dolphin/integration/IntegrationConstants.java]: hint

File: integration-common/src/main/java/com/canoo/dolphin/integration/enterprise/EnterpriseTestConstants.java [hunk 2]
```diff
@@ -0,0 +1,9 @@
+package com.canoo.dolphin.integration.enterprise;
+
+/**
```
Comment (timo0) [inline integration-common/src/main/java/com/canoo/dolphin/integration/enterprise/EnterpriseTestConstants.java]: hint

**Comment by hendrikebbers** [comment]
https://github.com/canoo/dolphin-platform/issues/297
https://github.com/canoo/dolphin-platform/issues/286

File: integration-tests/src/test/java/com/canoo/dolphin/integration/AbstractIntegrationTest.java [hunk 3]
```diff
@@ -45,12 +40,12 @@ protected void waitUntilServerIsUp(String host, long time, TimeUnit timeUnit) th
...
             }
             try {
-                URL healthUrl = new URL(host + "/health");
+                URL healthUrl = new URL(host + "/dolphin");
                 URLConnection connection = healthUrl.openConnection();
                 if(connection instanceof HttpURLConnection) {
                     HttpURLConnection httpURLConnection = (HttpURLConnection) connection;
                     httpURLConnection.connect();
-                    if(httpURLConnection.getResponseCode() == HttpStatus.HTTP_OK) {
+                    if(httpURLConnection.getResponseCode() == 405) {
```
Comment (timo0) [inline integration-tests/src/test/java/com/canoo/dolphin/integration/AbstractIntegrationTest.java]: Method Not Allowed?

Comment (hendrikebbers) [inline integration-tests/src/test/java/com/canoo/dolphin/integration/AbstractIntegrationTest.java]: yes, WIP ;)

**Comment by hendrikebbers** [comment]
@timo0 But please review the current state 👍

File: integration-tests-jee/src/main/webapp/WEB-INF/beans.xml [hunk 4]
```diff
@@ -0,0 +1 @@
+<beans/>
```
Comment (timo0) [inline integration-tests-jee/src/main/webapp/WEB-INF/beans.xml]: is this file needed?

**Comment by timo0** [comment]
We should discuss this next week

**Comment by timo0** [comment]
Please don't be shy and add detailed documentation in both the code and the documentation

File: build.gradle [hunk 5]
```diff
@@ -96,6 +96,8 @@ subprojects { subprj ->
         jcenter()
     }
 
+    test.useTestNG()
+
```
Comment (hendrikebbers) [inline build.gradle]: yes!
Later for some projects JUnit is configured. With the old setup, both TestNG & Junit were defined in an if-statement. Since JUnit is the default in Gravle, some project that should use TestNG were not configured to use TestNG

File: integration-common/src/main/java/com/canoo/dolphin/integration/action/ActionTestConstants.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
+package com.canoo.dolphin.integration.action;
+
+public interface ActionTestConstants {
+
+    String ACTION_CONTROLLER_NAME = "ActionController";
```
Comment (hendrikebbers) [inline integration-common/src/main/java/com/canoo/dolphin/integration/action/ActionTestConstants.java]: If you define a field in an interface it is by default `public static`

File: integration-tests-jee/src/main/webapp/WEB-INF/beans.xml (hunk 4 above)
Comment (hendrikebbers) [inline integration-tests-jee/src/main/webapp/WEB-INF/beans.xml]: Yes as it is defined in the CDI 1.0 spec. Can be removed with CDI 2.0 JavaEE 8 as far as I know. So maybe in 2 years or so 😞

**Comment by hendrikebbers** [comment]
@timo0 Once it is working I will add documentation.

File: integration-tests-jee/src/main/webapp/WEB-INF/beans.xml (hunk 4 above)
Comment (timo0) [inline integration-tests-jee/src/main/webapp/WEB-INF/beans.xml]: yay nice

File: integration-common/src/main/java/com/canoo/dolphin/integration/action/ActionTestConstants.java (hunk 6 above)
Comment (timo0) [inline integration-common/src/main/java/com/canoo/dolphin/integration/action/ActionTestConstants.java]: 👍

**Comment by hendrikebbers** [comment]
@aalmiray I do not know how I can change the build order. Maybe you can have a look at `integration-tests.gradle` in this branch?

What I want:

```
:integration-tests-jee:war
:integration-tests:copyWar
:integration-tests:compileTestJava
:integration-tests:composeUp
:integration-tests:test
:integration-tests:composeDown
```

Currently it looks like this:

```
:integration-tests:compileTestJava
:integration-tests:composeUp
:integration-tests:test
:integration-tests-jee:war
:integration-tests:copyWar
:integration-tests:composeDown
```

Docker will start before the WAR is created and copied. Tried several constellations but do not get it running. Can you have a look?

**Comment by aalmiray** [comment]
Define a task dependency between `composeUp' and 'copyWar'.
https://docs.gradle.org/3.4.1/userguide/tutorial_using_tasks.html#intro&gsc.tab=0

**Comment by hendrikebbers** [comment]
Tried it already:

`copyWar.isRequiredBy(dockerCompose)`

ends in:
`> Could not find method isRequiredBy() for arguments [com.avast.gradle.dockercompose.ComposeExtension_Decorated@30551308] on task ':integration-tests:copyWar' of type org.gradle.api.DefaultTask.
`

**Comment by hendrikebbers** [comment]
Or:

`dockerCompose.dependsOn copyWar`

ends in 
`> Could not find method dependsOn() for arguments [task ':integration-tests:copyWar'] on object of type com.avast.gradle.dockercompose.ComposeExtension.
`

**Comment by hendrikebbers** [comment]
Ahhh, it's
`composeUp.dependsOn copyWar`

:)

But why is `dockerCompose.isRequiredBy(test)` working? What is the difference?

**Comment by hendrikebbers** [comment]
@aalmiray build is working, THX

**Comment by hendrikebbers** [comment]
Now everything works fine on my system but on travis it looks like the WAR is not copied:

```
:integration-tests:copyWar
:integration-tests:composeUpBuilding paraya-integration
Step 1 : FROM airhacks/payara-configured
latest: Pulling from airhacks/payara-configured
Digest: sha256:29fa981a1abbf065c78f034b8fcccd6b612465395a69ba3f83be1dfd1b24c153
Status: Downloaded newer image for airhacks/payara-configured:latest
 ---> c075108fdf74
Step 2 : MAINTAINER Hendrik Ebbers, canoo.com
 ---> Running in f6d273dd2e84
 ---> 959547b77429
Removing intermediate container f6d273dd2e84
Step 3 : ADD integration-tests.war ${DEPLOYMENT_DIR}
Service 'paraya-integration' failed to build: lstat integration-tests.war: no such file or directory
:integration-tests:composeUp FAILED
```

@aalmiray @madmas any idea? Need help here :(

File: .gitignore [hunk 7]
```diff
@@ -11,3 +11,6 @@ node_modules/
 local.properties
 **/.nb-gradle
 
+integration-tests/docker-base/payara-integration/integration-tests.war
```
Comment (timo0) [inline .gitignore]: should we ignore *.war in general? or at least integration-tests.war?

File: integration-common/src/main/java/com/canoo/dolphin/integration/IntegrationConstants.java (hunk 1 above)
Comment (timo0) [inline integration-common/src/main/java/com/canoo/dolphin/integration/IntegrationConstants.java]: still there

File: .gitignore (hunk 7 above)
Comment (hendrikebbers) [inline .gitignore]: Done

**Comment by madmas** [comment]
@hendrikebbers done :)

**Comment by hendrikebbers** [comment]
@madmas :-*
You need to explain why this makes a change...
